resolver/resolver.go

```diff
@@ -1709,9 +1711,21 @@ func (r *Resolver) resolveMessageArgument(ctx *context, files []*File) {
 		msgArg := newMessageArgument(root.Message)
 		msgArg.Fields = append(msgArg.Fields, reqMsg.Fields...)
 		r.cachedMessageMap[msgArg.FQDN()] = msgArg
-		msgMap := make(map[string]*Message)
-		for _, msg := range r.resolveMessageArgumentRecursive(ctx, root) {
-			msgMap[msg.FQDN()] = msg
+		// Store the messages to serviceMsgMap first to avoid inserting duplicated ones to Service.Messages
+		for _, svc := range root.Message.File.Services {
+			if _, exists := serviceMsgMap[svc.FQDN()]; !exists {
+				serviceMsgMap[svc.FQDN()] = make(map[string]*Message)
+			}
+			for _, msg := range r.resolveMessageArgumentRecursive(ctx, root) {
+				serviceMsgMap[svc.FQDN()][msg.FQDN()] = msg
+			}
+		}
+	}
+
+	for fqdn, svc := range r.cachedServiceMap {
```

Comment:
If you change type of `serviceMsgMap` to `map[*Service]map[string]*Message` or `map[*Service]map[*Message]struct{}` , you can use `serviceMsgMap` here instead of `r.cachedServiceMap` .

Pointer's value to the same service or message are guaranteed to be the same.